✨clusterctl: validate provider metadata

kahirokunn opened the pull request:
This will enable early detection of #12241 such as these.

This PR adds stricter validation and richer diagnostics when reading provider metadata:

- Add full metadata.yaml content dump at `--v=5` log level to simplify debugging
- Implement `validateMetadata()` function to verify:
  - apiVersion matches clusterctl's expected group/version
  - kind is "Metadata"
  - releaseSeries is not empty
- Return clear, wrapped errors when decoding or validation fails
- Add comprehensive test cases for metadata validation

These changes help detect malformed metadata early and make troubleshooting easier by providing more detailed error messages and debug information.

fabriziopandini commented on cmd/clusterctl/client/repository/metadata_client.go:
```diff
@@ -84,6 +84,9 @@ func (f *metadataClient) Get(ctx context.Context) (*clusterctlv1.Metadata, error
 	}
 
+	// Dump the metadata.yaml content to logs for debugging
+	log.V(5).Info("metadata.yaml content", "provider", f.provider.ManifestLabel(), "version", f.version, "content", string(file))
```
I think we should not log config from the metadata file (generally we are not logging content from config files)

Looking at this from another angle, those kind of errors will mostrly hit users setting up a provider repository, and thus they already know the content of the file.

If those errors hits users, and the provider repository is not set up properly, there is not much they can do even if the log reports the full content of the file

kahirokunn replied:
That's right. I deleted it.

fabriziopandini reviewed (commented):
If there will be agreement on a stricter validation for the metadata file this a breaking change that should be documented in https://main.cluster-api.sigs.k8s.io/developer/providers/migrations/v1.10-to-v1.11.

I personally don't have strong opinions, might be interesting to understand better what use case lead to opening this PR

Also, probably worth checking if more clarification are required in https://main.cluster-api.sigs.k8s.io/developer/providers/contracts/clusterctl#metadata-yaml

kahirokunn commented:
Thanks for the thoughtful review!

The motivation for this PR came from the hours I spent tracking down [kubernetes-sigs/cluster-api#12241](https://github.com/kubernetes-sigs/cluster-api/issues/12241). I’m hoping to spare other users.

fabriziopandini commented on cmd/clusterctl/client/repository/metadata_client.go:
```diff
@@ -92,7 +92,39 @@ func (f *metadataClient) Get(ctx context.Context) (*clusterctlv1.Metadata, error
+func validateMetadata(metadata *clusterctlv1.Metadata, providerLabel string) error {
+	// Check if metadata has the correct apiVersion and kind
+	if metadata.APIVersion != clusterctlv1.GroupVersion.String() {
+		return errors.Errorf("invalid provider metadata: unexpected apiVersion %q for provider %s (expected %q). Full metadata.yaml content is available in logs with --v=5",
```
```suggestion
		return errors.Errorf("invalid provider metadata: unexpected apiVersion %q for provider %s (expected %q).",
```
This is not true anymore (same below)

kahirokunn replied:
Thank you. I fixed.

kahirokunn commented:
@fabriziopandini Thank you for the review! I've addressed your feedback and added the requested documentation:

## Changes made:

### 1. ✅ Added validation rules documentation to clusterctl contract
- Updated `docs/book/src/developer/providers/contracts/clusterctl.md`
- Added a new "Validation Rules" subsection under "Metadata YAML"
- Documented all three validation rules:
  - `apiVersion` must be `clusterctl.cluster.x-k8s.io/v1alpha3`
  - `kind` must be `Metadata`
  - `releaseSeries` must contain at least one entry
- Included error message examples for each validation rule

### 2. ✅ Added note about this change to migration guide
- Updated `docs/book/src/developer/providers/migrations/v1.10-to-v1.11.md`
- Created a new "clusterctl" section as requested
- Explained the stricter validation for provider metadata
- Clarified that providers with invalid metadata.yaml files will need to update them

The documentation now clearly explains the validation rules and helps providers understand what changes they might need to make to their metadata.yaml files.

Please let me know if you'd like any adjustments to the documentation!

fabriziopandini reviewed (commented):
/lgtm

cc @sbueringer

sbueringer commented on cmd/clusterctl/client/repository/metadata_client.go:
```diff
@@ -92,7 +92,39 @@ func (f *metadataClient) Get(ctx context.Context) (*clusterctlv1.Metadata, error
 	}
 
-	//TODO: consider if to add metadata validation (TBD)
+	if err != nil {
```
This if seems like unreachable code. The error is already checked above

kahirokunn replied:
Thx. I fixed it.

sbueringer commented on docs/book/src/developer/providers/contracts/clusterctl.md:
```diff
@@ -201,6 +201,25 @@ releaseSeries:
+
+1. **apiVersion**: Must be set to `clusterctl.cluster.x-k8s.io/v1alpha3`
+   * This ensures compatibility with the current clusterctl metadata format
+   * Invalid apiVersion will result in an error: `invalid provider metadata: unexpected apiVersion "<value>" for provider <provider-name> (expected "clusterctl.cluster.x-k8s.io/v1alpha3")`
```
I would not document the exact error messages here.

1. There's a very low chance we'll keep these up-to-date when the code changes
2. The exact wording of the error message should not be relevant for users
3. The error message format is not part of our API / contract

kahirokunn replied:
Thx. I fixed.

sbueringer commented on docs/book/src/developer/providers/contracts/clusterctl.md:
```diff
@@ -201,6 +201,25 @@ releaseSeries:
+   * Empty releaseSeries will result in an error: `invalid provider metadata: releaseSeries is empty in metadata.yaml for provider <provider-name>`
+
+These validation rules help catch configuration issues early and provide clear error messages to assist in troubleshooting.
+Ask Copilot about this file-diff
```
Copilot? Please remove

kahirokunn replied:
Sorry, I deleted it.

fabriziopandini reviewed (commented):
/lgtm
